hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/CryptoInputStream.java

```diff
@@ -341,6 +343,26 @@ public int read(long position, byte[] buffer, int offset, int length)
           "positioned read.");
     }
   }
+
+   /** Positioned read using ByteBuffers. It is thread-safe */
+  @Override
+  public int read(long position, final ByteBuffer buf)
+      throws IOException {
+    checkStream();
+    try {
+      int pos = buf.position();
+      final int n = ((ByteBufferPositionedReadable) in).read(position, buf);
+      if (n > 0) {
+        // This operation does not change the current offset of the file
+        decrypt(position, buf, n, pos);
+      }
+
+      return n;
+    } catch (ClassCastException e) {
```

Comment:
Here we catch the ClassCastException to check whether the inpustream has implemented an ByteBufferPositionedReadable interface ?  I don't think it's a good way, because the catch{...} block need an full stack and not so efficient .  
How about use the following : 

```java
if(in instanceof ByteBufferPositionedReadable){
    //...
}else{
    throw new UnsupportedOperationException(...)
}
```

Reply:
Yeah, its probably not the most efficient way to do things, but thats how all other methods handle the same thing: `PositionedReadable`, `Seekable`, etc. + the exception handling code wouldn't be on the hot path. So in this case I would prefer to keep the code consistent with the rest of the class.

Reply:
add an instance check; all the FSDataInputStream classes do that

Reply:
Ok, makes sense will do. Any objections if I just fix this here rather than filing a separate JIRA? The changes are pretty small / it would probably be less effort than creating, reviewing, and merging a separate patch just for this.

Reply:
Fixed